include/ccmath/math/fmanip/ldexp.hpp

```diff
@@ -8,7 +8,69 @@
 
 #pragma once
 
+#include <ccmath/internal/support/bits.hpp>
+#include "ccmath/internal/support/floating_point_traits.hpp" //this is included in bits, but i'm going to include it anyways
+
 namespace ccm
 {
+	// this function multiples parameter x by two to the power of paramter powerOf2
+	// x * 2^powerOf2
+	template <typename T, std::enable_if_t<!std::is_integral_v<T>, bool> = true>
+	inline constexpr T ldexp(T x, int32_t powerOf2) noexcept
+	{
+		if (!ccm::isfinite(x)) { return x; }
+		int32_t oldexp = ccm::support::get_exponent_of_floating_point<T>(x);
+
+		// if the mantissa is 0 and the original exponent is 0
+		if ((oldexp == 0) && ((ccm::support::bit_cast<ccm::helpers::float_bits_t<T>>(x) &
+							   ccm::helpers::floating_point_traits<T>::normal_mantissa_mask) == 0))
+		{
+			return x;
+		}
+
+		if (powerOf2 > ccm::helpers::floating_point_traits<T>::maximum_binary_exponent)
+		{
+			// error == hugeval
+			return std::numeric_limits<T>::infinity();
+		}
+		// the reference source says -2 * exp_max
+		else if (powerOf2 < ccm::helpers::floating_point_traits<T>::minimum_binary_exponent)
+		{
+			// error == range
+			return 0;
+		}
+		// normalizes an abnormal floating point
+		if (oldexp == 0)
+		{
+			x *= ccm::helpers::floating_point_traits<T>::normalize_factor;
+			powerOf2 = -sizeof(T) * 8; //8 is bits in a byte
+			oldexp	 = ccm::support::get_exponent_of_floating_point<T>(x);
+		}
+
+		powerOf2 = oldexp + powerOf2;
+		if (powerOf2 >= ccm::helpers::floating_point_traits<T>::maximum_binary_exponent)
+		{
+			// overflow
+			return std::numeric_limits<T>::infinity();
+		}
+		if (powerOf2 > 0)
+		{
+			return ccm::support::set_exponent_of_floating_point<T>(x, powerOf2);
+		}
+		// denormal, or underflow
+		powerOf2 += sizeof(T) * 8; //8 is bits in a byte
+		x = ccm::support::set_exponent_of_floating_point<T>(x, powerOf2);
+		x /= ccm::helpers::floating_point_traits<T>::normalize_factor;
+		if (x == static_cast<T>(0))
+		{
+			// underflow report
+		}
+		return x;
+	}
 
+	template <typename T, std::enable_if_t<std::is_integral_v<T>, bool> = true>
+	inline constexpr double ldexp(T x, int32_t powerOf2)
+	{
+		return ldexp(static_cast<double>(x), powerOf2);
+	}
 } // namespace ccm
```

Comment:
Missing the specializations of ldexp that are defined by the standard.

```cpp
float       ldexpf( float num, int exp );

long double ldexpl( long double num, int exp );
```

Reply:
Have you figured out how you want to implement long double yet?

Reply:
Currently, if a proper implementation of long double is possible (an example of this is `ccm::scalbnl`) then we should strive to implement it. Though in  cases where implementing a long double version is not a simple manner (e.g. in instances were bit manipulation is required) I am currently allowing the implementer to cast the long double to double and to use the double implementation instead. (You can see examples of me doing this in `ccm::exp2`)